add article about redux (#8)
Slashgear/slashgear.github.io: src/pages/react-redux-pitfalls-and-best-pratices/index.md

### Do not create a reference in the _selector_

Let's imagine the next _selector_:

```js
const getUserById = userId => state =>
  state.users.find(user => user.id === userId) || {}
```

The developer here wanted to ensure that his _selector_ is null safe and always returns an _object_.
This is something we see quite often.

Each time this selector will be called for a `user` not present in the state, it will return a new object, a new reference.

> With [useSelector](), returning a new object every time will always force a re-render by default.
> [Doc of react-redux](https://react-redux.js.org/api/hooks#equality-comparisons-and-updates)

To counter this, you can try to store a unique reference to this default value to avoid changing it.

```js
const defaultUser = {}

const getUserById = userId => state =>
  state.users.find(user => user.id === userId) || defaultUser
```

The same goes for the selector usage that returns a new ref at each call.
The use of the `filter` function returns a new array each time a new reference even if the filter conditions have not changed.

A selector should not return a _view_ (a copy) of the state but directly what it contains.

> By respecting this principle, your components will return only if an action modifies the state.

Reviewer: ```suggestion
By respecting this principle, your components will rerender only if an action modifies the state.
```

> At Bedrock, we avoid these problems by not storing in the state the information as we want to display it.

Reviewer: No sure of this one 🤔

What I recommend is a functional splitting of the application.
Your file structure should represent the business more than your technical layers.
Some very good articles explain it notably through the use of [DDD principles](https://en.wikipedia.org/wiki/Domain-driven_design).

In Bedrock, we use a folder named _modules_ which groups together the different folders associated with the feature of your application.

```txt
app/
  modules/
    user/
      __tests__/
        user.reducer.spec.js
      components/
      user.reducer.js
    product/
      __tests__/
        product.reducer.spec.js
      components/
      product.reducer.js
    account/
      __tests__/
      account.reducer.spec.js
      components/
      account.reducer.js
  store.js
  index.js
```

So in `store.js` all you have to do is combine your different reducers.

```js
import { createStore, combineReducers } from 'redux'
import { user } from './modules/user/user.reducer.js'
import { product } from './modules/user/product.reducer.js'
import { account } from './modules/user/account.reducer.js'

export const store = createStore(combineReducers({ user, product, account }))
```

By following this principle, you can:

> - keep reducers readable because they have a limited scope.

Reviewer 2: ```suggestion
- keep reducers readable because they have a limited scope
```

- structure and define the functionalities of your application
- facilitate the testing of your reducers

> Historically, this segmentation has allowed us to remove complete application areas without having imapcts on the entire codebase, just by deleting the `module` folder associated with the feature.

Reviewer 2: ```suggestion
Historically, this segmentation has allowed us to remove complete application areas without impacting the entire codebase, just by deleting the `module` folder associated with the feature.
```

### Proxy access to the state

Now that your reducers have been placed in the functional `module`, you need to allow your components to access the state via `selector`.
A `selector` is a function that has the `state` as a parameter, and retrieves its information.

```js
export const getUserName = ({ user: { lastName } }) => lastName
```

You can also pass parameters to a `selector` by wrapping it with a function.

```js
export const getProduct = productId => ({ product: { list } }) =>
  list.find(product => product.id === productId)
```

This will allow you to use them in your components using the [useSelector] hook.

```js
const MyComponent = () => {
  const product = useSelector(getProduct(12))
  return <div>{product.name}</div>
}
```

It is specified in the `react-redux` doc that the _selector_ is called for each render of the component.

> If the `selector` function reference does not change, a cached version of the object can be returned directly.

Author: Et bien non selon la doc => https://react-redux.js.org/api/hooks#equality-comparisons-and-updates

> If the `selector` function reference does not change, a cached version of the object can be returned directly.

Reviewer 3: oh oui! J'étais sur le mapStateToProps encore, désolé 😄

```txt
app/
  modules/
    user/
      __tests__/
        user.reducer.spec.js
      components/
      user.reducer.js
      user.selectors.js <--- This is where all module selectors are exported
```

### Prefix the name of your actions

> I really advise you to define naming rules for your actions and if possible check them with an `eslint` rule.

> Actions are in uppercase letters separated by '\_'.

Reviewer 2: ```suggestion
Actions are in uppercase letters separated by '_'.
```

> Exemple avec cette action: `SET_USERS`.

Reviewer 2: ```suggestion
Here an example with this action: `SET_USERS`.
```

```txt
app/
  modules/
    user/
      __tests__/
        user.reducer.spec.js
      components/

>       user.actions.js <--- This is where all module action creator are exported

Reviewer 2: ```suggestion
      user.actions.js <--- This is where all module action creators are exported
```

      user.reducer.js
      user.selectors.js
```

Action names are prefixed by the name of the `module` in which it is located.
This gives a full name: `user/SET_USERS`.
A big advantage of this naming rule is that you can easily filter the action in [redux-devtools](https://github.com/reduxjs/redux-devtools).

![redux devtools screenshot](./redux-devtools.png)

### Always test your reducers

> The `reducers` are the holders of your application's business.

Author: Alors si on suit le StyleGuide de redux à la règle, c'est le reducer qui doit porter les logiques. 

J'ai un avis plus mitigé.
Je pense que les actions/middleware sont la pour exprimer les _effets_ et le reducer représente le _domaine_

> The `reducers` are the holders of your application's business.

Reviewer 3: Je viens de voir que la doc de redux pointe vers un article qui décortique chacune des manières de gérers la business logique => https://medium.com/@jeffbski/where-do-i-put-my-business-logic-in-a-react-redux-application-9253ef91ce1 . Et ils mentionne un peu toutes les solutions avec les + and - (au cas ou)

> The `reducers` are the holders of your application's business.

Author: Vraiment cool, je vais me lire ça ce midi

They manipulate the state of your application.

**This code is therefore _sensitive_.**

> ➡️ a modification can have a lot of impact on your application.

Reviewer 2: ```suggestion
➡️ A modification can have a lot of impact on your application.
```

**This code is rich in business rules**

➡️ You must be ensured that these are correctly implemented.

The good news is that this code is relatively easy to test.
A [reducer] is a single function that takes 2 parameters.

> This function will return a new `state` depending on the type of action and these parameters.

Reviewer 2: ```suggestion
This function will return a new `state` depending on the type of action and its parameters.
```

This is the standard structure for testing [reducer]s with [Jest](https://jestjs.io/):

```js
describe('ReducerName', () => {
  beforeEach(() => {
    // Init a new state
  })
  describe('ACTION', () => {
    // Group tests by action type
    it('should test action with some params', () => {})
    it('should test action with other params', () => {})
  })
  describe('SECOND_ACTION', () => {
    it('should test action with some params', () => {})
  })
})
```

I also recommend that you use the [deep-freeze](https://www.npmjs.com/package/deep-freeze) package on your `state` to ensure that all actions return new references.

> Ultimately, testing your [reducer] will allow you to easily refactor the internal structure of their state without the risk of introducing regressions.

Reviewer 2: ```suggestion
Ultimately, testing your [reducer]s will allow you to easily refactor the internal structure of their state without the risk of introducing regressions.
```

## Keep the immutability and readability of your reducers

A [reducer] is a function that must return a new version of the state containing its new values while keeping the same references of the objects that have not changed.

> This allows you to take full advantage of _Structural sharing_ and avoid exploding your memory usage.

Author: On a pas eu de souci, mais j'ai déjà vu des appli se crouter à force de faire des deepClone x)

The use of the _spread operator_ is thus more than recommended.

However, in the case where the state has a complicated and deep structure, it can be verbose to change the state without destroying the references that should not change.

For example, here we want to override the `a.y.A` value of the state while keeping the objects that don't change.

```js
const state = {
  a: {
    z: {
      A: 'A',
    },
    y: {
      A: 'A',
    },
  },
  b: {
    z: {
      A: 'A',
    },
  },

> }

Author: Good Idea

> // When you want tu change nested state value and use immutability

Reviewer 2: ```suggestion
// When you want to change nested state value and use immutability
```

const newState = {
  ...state,
  a: {
    ...state.z,
    y: {
      A: 'B',
    },
  },
}
```

> To avoid this, [a member of the Bedrock team](https://github.com/flepretre) released a package that allows `set` one nested attribute while ensuring immutability: [immutable-set]

Reviewer 3: ~~one~~ => nested attributes

> This package is much easier to use than tools like [immutable.js].

Reviewer 3: why? pas super fair de tackler les autres outils sans montrer la différence :p

```js
import set from 'immutable-set'

const newState = set(state, 'a.y.A', 'B')
```

## Do not use the default case

The implementation of a `redux` [reducer] very often consists of a `switch` where each `case` corresponds to an `action`.
A `switch` must always define the `default` case if you follow so basic `eslint` rules.

Let's imagine the following [reducer]:

```js
const initialState = {
  value: 'bar',
  index: 0,
}

function reducer(initialState, action) {
  switch (action.type) {
    case 'FOO':
      return {
        value: 'foo',
      }
    default:
      return {
        value: 'bar',
      }
  }
}
```

We can naively say that this [reducer] manages two different actions. It's okay.
If we isolate this [reducer] there are only two types of `action' that can change this state; the`FOO' action and any other action.

However, if you have followed the advice to cut out your reducers, you don't have only one reducer acting on your blind.

That's where the previous [reducer] is a problem.
Indeed, any other action will change this state to a `default` state.

> A `dispatch` action will pass through each of the reducers associated with this one.

Reviewer 3: Une image serait tellement cool pour montrer ca! Beaucoup de pesonnes l'oublies! (et je l'avais moi même oublié en essayant de bosser avec les sagas la premiere fois)

This should be avoided.

If you want to modify the state with an action from another module, you can do so by adding a `case` on that action.

```js
function reducer(state = initialState, action) {
  switch (action.type) {
    case 'FOO':
      return {
        value: 'foo',
      }
    case 'otherModule/BAR':
      return {
        value: 'bar',
      }
    default:
      return state
  }
}
```

## Use custom middlewares

I've often seen `action` behaviors being copied and pasted, from action to action.
When you're a developer, "copy-paste" is never the right way.

The most common example is handling HTTP calls during an action that uses `redux-thunk`.

```js
export const foo = () =>
  fetch('https://example.com/api/foo')
    .then(data => ({ type: 'FOO', data }))
    .catch(error => {
      // Do something
    })

export const bar = () =>
  fetch('https://example.com/api/bar')
    .then(data => ({ type: 'BAR', data }))
    .catch(error => {
      // Do something
    })
```

These two actions are basically the same thing, we could very well make a factory that would do the code in common.

> If this pattern repeated itself only twice, that would probably be the best thing to do.

Reviewer 3: Je suis pas sur que ce que tu essaies de dire soit clair pour tout le monde 😅

> If this pattern repeated itself only twice, that would probably be the best thing to do.

Reviewer 2: ```suggestion
If this pattern is repeated itself only twice, that would probably be the best thing to do.
```

> Basically the _meta_ action we want to represent here when it is `dispatch`:

Reviewer 3: `dispatch`ed

> Basically the _meta_ action we want to represent here when it is `dispatch`:

Reviewer 2: ```suggestion
Basically the _meta_ action we want to represent here when it is dispatched:
```

```
Fetch something
-- return action with the result
-- in case or error, do something
```

We could very well define a middleware that would take care of this behavior.

```js
const http = store => next => async action => {
  if (action.http) {
    try {
      action.result = await fetch(action.http)
    } catch (error) {
      // Do something
    }
  }
  return next(action)
}

// in redux store init
const exampleApp = combineReducers(reducers)
const store = createStore(exampleApp, applyMiddleware(http))
```

> Thus the two preceding actions could be written much more simply:

Reviewer 3: On comprend, mais `simpler` est le comparatif adapté plutot que more simply je crois

```js
export const foo = () => ({ type: 'FOO', http: 'https://example.com/api/foo' })

export const bar = () => ({ type: 'BAR', http: 'https://example.com/api/bar' })
```

The big advantages of using middleware in a complex application:

> - Avoids code duplication

Reviewer 2: For consistency with previous lists 😇 
```suggestion
- avoids code duplication
```

> - Allows you to define common behaviors between your actions

Reviewer 2: ```suggestion
- allows you to define common behaviors between your actions
```

> - Standardize redux _meta_ action types

Reviewer 2: ```suggestion
- standardize redux _meta_ action types
```

## Avoid redux related rerender

The trick when using redux is to trigger component re-render when you connect them to the state.

> Even if [rerenders are not always a problem], re-render caused by the use of redux are really to be prevented.

Reviewer 2: ```suggestion
Even if [rerenders are not always a problem], re-render caused by the use of redux really has to be prevented.
```

Just beware of the following traps.

### Do not create a reference in the _selector_

Let's imagine the next _selector_:

```js
const getUserById = userId => state =>
  state.users.find(user => user.id === userId) || {}
```

> The developer here wanted to ensure that his _selector_ is null safe and always returns an _object_.

Reviewer 3: Pour les problématiques de genre du monde actuelle: `his` => `its / their`

> [Doc of react-redux](https://react-redux.js.org/api/hooks#equality-comparisons-and-updates)

However in the case of an object, as in the example above (or an array), the reference of this default value is new each time the selector is executed.
Similarly for the default values in destructuring, you should never do this :

```js
const getUsers = () => ({ users: [] }) => users
```

What to do then?
Whenever possible, the default values should be stored in the reducer.
Otherwise, the default value must be extracted into a constant so that the reference remains the same.

```js
const defaultUser = {}

const getUserById = userId => state =>
  state.users.find(user => user.id === userId) || defaultUser
```

The same goes for the selector usage that returns a new ref at each call.
The use of the `filter` function returns a new array each time a new reference even if the filter conditions have not changed.

To continue, it is important that [useSelector] does not return a function.
Basically you should never do this:

```js
const getUserById = state => userId =>
  state.users.find(user => user.id === userId)
const uider = useSelector(getUserById)(userId)

> ```

Reviewer 3: Est ce qu'on est sur de ca? Parce qu'il y a deux implications dans le selector:

- le memoize du return du selector dans le cas ou le state est le même
- le rerender à cause du memo / pure components

Mais je me dis que si le sub state (reselect) ne change pas en référence, la valeur mémoizé ne va pas changer en référence. Et donc on retourner quelque chose qui aura executer une seule fois le filter, et qui sera le même en référence (puisque mémoizé) ?

Reviewer 3: Oops, j'avais pas encore lu le point suivant 😅

We would quickly tend to generate it in the component directly.

```js
const MyComponent = () => {
  const user = useSelector(getUser)

  return (
    <div>
      <h1>{user.name}</h1>
      <img src={`https://profil-pic.com/${user.id}`} />
    </div>
  )
}
```

Here, the url of the image is dynamically computed in the component, and thus at each render.
We prefer to modify our reducers in order to include a `profileUrl` attribute so that this information is directly accessible.

```js
switch (action.type) {
  case `user/SET_USER`:
    return {
      ...state,
      user: {
        ...action.user,
        profilUrl: `https://profil-pic.com/${action.user.id}`,
      },
    }
}
```

This information is then calculated once per action and not every time it is rendered.

## Don't use _useReducer_ for your business data

Since the arrival of hooks, we have many more tools provided directly by React to manage the state of our components.
The [useReducer] hook allows to set a state that can be modified through actions.
We're really very very close to a redux state that we can associate to a component, it's great.

> However, if you use redux in your application, it seems quite strange to have to use use useReducer.

Reviewer 2: ```suggestion
However, if you use redux in your application, it seems quite strange to have to use useReducer.
```

---
title: 'React/Redux: pitfalls and best practices'
description:
  We use React and Redux for almost 4 years at Bedrock on our video platform named 6play.
  Good practices and mistakes to be avoided have been gathered in this article.
date: 2020-04-04
hero: ./redux.jpg
language: en
---

After 2 years using React with Redux for the video platform [6play](https://6play.fr), I was able to identify good practices and pitfalls to avoid at all costs.
The [Bedrock](https://www.bedrockstreaming.com/) team (to which I currently belong) kept the technical stack of the project up to date to take advantage of the new features of `react`, `react-redux` and `redux`.

So here are my tips for maintaining and using React and Redux in your application without going mad.

_This article is not an introduction to React or Redux. I recommend [this documentation](https://redux.js.org/basics/usage-with-react) if you want to see how to implement it in your applications._

You could also take a look at [Redux offical style guide] in which you could find some of those tips and others.

> ## Avoid having only one reducer

Reviewer 4: Ca me parait pas bien pertinent de publier ce "conseil" en 2020.

Author: Pour toi, c'est bien évident c'est sûr

The [reducer] is the function that is in charge of building a new state at each `action`.
One might be tempted to manipulate only one reducer.
In the case of a small application, this is not a problem.
For applications expressing a complex and evolving business, it is better to opt in for the [combineReducers] solution.

This feature of `redux` allows to manipulate not one but several [reducer]s which act respectively on the state.

> When and how to split its application?

What I recommend is a functional splitting of the application.
In my approach, I would tend to represent the business of the application more than the technical stuff implied.

> Some very good articles explain it notably through the use of [DDD principles](https://en.wikipedia.org/wiki/Domain-driven_design).

Reviewer 5: Maybe you could use the `We` (bedrock) instead of the `I`. It would give more weight to your arguments.
What do you think?

In Bedrock, we use a folder named _modules_ which groups together the different folders associated with the feature of your application.

```txt
app/
  modules/
    user/
      __tests__/
        user.reducer.spec.js
      components/
      user.reducer.js
    product/
      __tests__/
        product.reducer.spec.js
      components/
      product.reducer.js
    account/
      __tests__/
      account.reducer.spec.js
      components/
      account.reducer.js
  store.js
  index.js
```

So in `store.js` all you need to do is combine your different reducers.

```js
import { createStore, combineReducers } from 'redux'
import { user } from './modules/user/user.reducer.js'
import { product } from './modules/user/product.reducer.js'
import { account } from './modules/user/account.reducer.js'

export const store = createStore(combineReducers({ user, product, account }))
```

> By following this principle, you can:

Reviewer 5: `you can` => `you will`
i would translate that in : `Tu peux / Vous pouvez` => `Tu vas / Vous allez`.
What do you think?

This allows you to take full advantage of _Structural sharing_ and avoid exploding your memory usage.
The use of the _spread operator_ is thus more than recommended.

However, in the case where the state has a complicated and deep structure, it can be verbose to change the state without destroying the references that should not change.

For example, here we want to override the `Rhone.Villeurbanne.postal` value of the state while keeping the objects that don't change.

```js
const state = {
  Rhone: {
    Lyon: {
      postal: '69000' ,
    },
    Villeurbanne: {
      postal: '',
    },
  },
  Isère: {
    Grenoble: {
      postal: '39000',
    },
  },
}

// When you want to change nested state value and use immutability
const newState = {
  ...state,
  Rhone: {
    ...state.Lyon,
    Villeurbanne: {
      postal: '69100',
    },
  },
}
```

To avoid this, [a member of the Bedrock team](https://github.com/flepretre) released a package that allows to `set` nested attribute while ensuring immutability: [immutable-set]

> This package is much easier to use than tools like [immutable.js] because it does not use Object prototype.

Reviewer 4: Il manque des liens là non ? ou tu veux vraiment afficher le nom des libs entre crochets? 🤔

```js
import set from 'immutable-set'

const newState = set(state, `Rhone.Villeurbanne.postal`, '69100')
```

> ## Do not use the default case

Reviewer 4: Ca me viendrait pas à l'idée. Personne fait ça non?

Author: Malheureusement, un jour je l'ai trouvé dans un reducer. 
Ma tristesse n'avait pas d'égal

We would quickly tend to generate it in the component directly.

```js
const MyComponent = () => {
  const user = useSelector(getUser)

  return (
    <div>
      <h1>{user.name}</h1>
      <img src={`https://profil-pic.com/${user.id}`} />
    </div>
  )
}
```

Here, the url of the image is dynamically computed in the component, and thus at each render.
We prefer to modify our reducers in order to include a `profileUrl` attribute so that this information is directly accessible.

```js
switch (action.type) {
  case `user/SET_USER`:
    return {
      ...state,
      user: {
        ...action.user,
        profilUrl: `https://profil-pic.com/${action.user.id}`,
      },
    }
}
```

This information is then calculated once per action and not every time it is rendered.

## Don't use _useReducer_ for your business data

Since the arrival of hooks, we have many more tools provided directly by React to manage the state of our components.
The [useReducer] hook allows to set a state that can be modified through actions.
We're really very very close to a redux state that we can associate to a component, it's great.

> However, if you use redux in your application, it seems quite strange to have to use useReducer.

Reviewer 5: useReducer should be in quote no?
